src/domain/task-system/services/src/task_scheduler_impl.rs

```diff
@@ -101,26 +130,43 @@ impl TaskScheduler for TaskSchedulerImpl {
         dataset_id: &DatasetID,
         pagination: TaskPaginationOpts,
     ) -> Result<TaskStateListing, ListTasksByDatasetError> {
-        let total_count = self
-            .event_store
-            .get_count_tasks_by_dataset(dataset_id)
-            .await?;
+        let total_count = <DatabaseTransactionRunner>::run_transactional_with(
+            &self.catalog,
+            |event_store: Arc<dyn TaskSystemEventStore>| async move {
+                let total_count = event_store.get_count_tasks_by_dataset(dataset_id).await?;
 
-        let dataset_id = dataset_id.clone();
+                Ok(total_count)
+            },
+        )
+        .await?;
 
         use futures::TryStreamExt;
+
+        let dataset_id = dataset_id.clone();
         let stream = Box::pin(async_stream::stream! {
-            let relevant_task_ids: Vec<_> = self
-                .event_store
-                .get_tasks_by_dataset(&dataset_id, pagination)
-                .await
-                .try_collect()
-                .await
-                .int_err()?;
+            let relevant_task_ids = DatabaseTransactionRunner::run_transactional_with(
+                &self.catalog,
+                |event_store: Arc<dyn TaskSystemEventStore>| async move {
+                    event_store
+                        .get_tasks_by_dataset(&dataset_id, pagination)
+                        .await
+                        .try_collect::<Vec<_>>().await
+                },
+            )
+            .await
+            .int_err()?;
 
             // TODO: implement batch loading
             for task_id in relevant_task_ids {
-                let task = Task::load(task_id, self.event_store.as_ref()).await.int_err()?;
+                let task = DatabaseTransactionRunner::run_transactional_with(
+                    &self.catalog,
+                    |event_store: Arc<dyn TaskSystemEventStore>| async move {
+                        Task::load(task_id, event_store.as_ref()).await
+                    },
+                )
+                .await
+                .int_err()?;
+
                 yield Ok(task.into());
```

Comment:
This looks very suspicious. Too many transactions just to iterate a list, and too many queries.
A simple possibility could be to load all tasks in one transactoins. This is fine if we are talking about few dozens.
But there might be more, so maybe we need chunks - load a chunk of tasks in 1 trip to the database, yield those while available in temporary collection, then proceed to the next chunk.